### PR TITLE
Add a script for enum type flags code generation 

### PR DESCRIPTION
**Pull request by s373r**
## Description

This PR adds a basic type flags generator for enumerations

---

Related issue: https://github.com/kamu-data/kamu-cli/issues/390
Code review comment: https://github.com/kamu-data/kamu-cli/pull/523#discussion_r1508918831

## Codegen result

**Note:** the following code after calling script `make codegen-odf-serde` from `kami-cli` repository 

```rust
// Copyright Kamu Data, Inc. and contributors. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

///////////////////////////////////////////////////////////////////////////////
// WARNING: This file is auto-generated from Open Data Fabric Schemas
// See: http://opendatafabric.org/
///////////////////////////////////////////////////////////////////////////////

use bitflags::bitflags;

use crate::MetadataEvent;

///////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////////////////////
// MetadataEvent
// https://github.com/kamu-data/open-data-fabric/blob/master/open-data-fabric.md#metadataevent-schema
////////////////////////////////////////////////////////////////////////////////

bitflags! {
    #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
    pub struct MetadataEventTypeFlags: u32 {
        const ADD_DATA = 1 << 0;
        const EXECUTE_TRANSFORM = 1 << 1;
        const SEED = 1 << 2;
        const SET_POLLING_SOURCE = 1 << 3;
        const SET_TRANSFORM = 1 << 4;
        const SET_VOCAB = 1 << 5;
        const SET_ATTACHMENTS = 1 << 6;
        const SET_INFO = 1 << 7;
        const SET_LICENSE = 1 << 8;
        const SET_DATA_SCHEMA = 1 << 9;
        const ADD_PUSH_SOURCE = 1 << 10;
        const DISABLE_PUSH_SOURCE = 1 << 11;
        const DISABLE_POLLING_SOURCE = 1 << 12;
    }
}

impl From<&MetadataEvent> for MetadataEventTypeFlags {
    fn from(v: &MetadataEvent) -> Self {
        match v {
            MetadataEvent::AddData(_) => Self::ADD_DATA,
            MetadataEvent::ExecuteTransform(_) => Self::EXECUTE_TRANSFORM,
            MetadataEvent::Seed(_) => Self::SEED,
            MetadataEvent::SetPollingSource(_) => Self::SET_POLLING_SOURCE,
            MetadataEvent::SetTransform(_) => Self::SET_TRANSFORM,
            MetadataEvent::SetVocab(_) => Self::SET_VOCAB,
            MetadataEvent::SetAttachments(_) => Self::SET_ATTACHMENTS,
            MetadataEvent::SetInfo(_) => Self::SET_INFO,
            MetadataEvent::SetLicense(_) => Self::SET_LICENSE,
            MetadataEvent::SetDataSchema(_) => Self::SET_DATA_SCHEMA,
            MetadataEvent::AddPushSource(_) => Self::ADD_PUSH_SOURCE,
            MetadataEvent::DisablePushSource(_) => Self::DISABLE_PUSH_SOURCE,
            MetadataEvent::DisablePollingSource(_) => Self::DISABLE_POLLING_SOURCE,
        }
    }
}
```

### CODE REVIEW

File: tools/jsonschema_to_rust_dto_enum_flags.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        if os.path.isdir(path):
+            read_schemas_rec(path, schemas)
+            continue
+
+        with open(path) as f:
+            s = json.load(f)
+            fname = os.path.splitext(os.path.split(path)[-1])[0]
+            name = os.path.splitext(s['$id'].split('/')[-1])[0]
+            assert fname == name, f"{fname} != {name}"
+            schemas[name] = s
```
Comment (sergiimk) [inline tools/jsonschema_to_rust_dto_enum_flags.py:81]: I feel like none of this complexity is actually required for this script - it only needs one single schema `MetadataEvent.json` and could just load it directly.

File: tools/jsonschema_to_rust_dto_enum_flags.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    yield 'bitflags! {'
+    yield from indent(['#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]'])
+    yield from indent([f'pub struct {enum_name}: u32 {{'])
+
+    enames = []
+
+    for i, isch in enumerate(sch["oneOf"]):
+        ref = cast(str, isch["$ref"])
+        ename = ref.split('/')[-1]
+        yield from indent(indent([f'const {to_screaming_snake_case(ename)} = 1 << {i};']))
```
Comment (sergiimk) [inline tools/jsonschema_to_rust_dto_enum_flags.py:111]: What will happen to `const DATA_BLOCK = Self::ADD_DATA.bits() | Self::EXECUTE_TRANSFORM.bits();`?

Comment (s373r) [reply tools/jsonschema_to_rust_dto_enum_flags.py:111]: This has already been added via `dtos_extra.rs`:

```rust
impl MetadataEventTypeFlags {
    pub const DATA_BLOCK: Self =
        Self::from_bits_retain(Self::ADD_DATA.bits() | Self::EXECUTE_TRANSFORM.bits());
}
```
(c) https://github.com/kamu-data/kamu-cli/pull/543/files#diff-b967f0a7157d57be5b442d9f8b58c5828e05b3e2efd589e0ef8bf30228550fe9R305-R308

File: tools/jsonschema_to_rust_dto_enum_flags.py (hunk 1 above)
Comment (s373r) [reply tools/jsonschema_to_rust_dto_enum_flags.py]: Simplified -- Done

File: tools/jsonschema_to_rust_dto_enum_flags.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        print(f'// {name}')
+        print('// ' + DOCS_URL.format(name.lower()))
+        print('/' * 80)
+        print()
+
+        for l in render_schema(name, sch):
+            print(l)
+        print()
+
+
```
Comment (sergiimk) [inline tools/jsonschema_to_rust_dto_enum_flags.py:45]: please remove two empty lines

(I know we should have a code auto-formatter, but I was hoping to rewrite these scripts to Rust for a long time not to invest time in Python CI)

File: tools/jsonschema_to_rust_dto_enum_flags.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    except Exception as ex:
+        raise Exception(
+            f'Error while rendering {name} schema:\n{sch}'
+        ) from ex
+
+
+def read_schema(schemas_dir: str, file_path: str) -> (SchemaName, JsonType):
```
Comment (sergiimk) [inline tools/jsonschema_to_rust_dto_enum_flags.py:52]: Can you change this to use `utils/schemas.py`? It has an identical function.

I started introducing common utils module, but did not yet update some codegens to use it. Sorry for the mess.

File: tools/jsonschema_to_rust_dto_enum_flags.py (hunk 3 above)
Comment (sergiimk) [reply tools/jsonschema_to_rust_dto_enum_flags.py:45]: in Rust we would even be able to use `syn` crate to generate code ... *sigh* I wish we had an extra year to dedicate fully to the tech debt.

**Review by sergiimk — APPROVED** [review]
Just a few minor points - feel free to merge when done :+1:

File: tools/jsonschema_to_rust_dto_enum_flags.py (hunk 3 above)
Comment (s373r) [reply tools/jsonschema_to_rust_dto_enum_flags.py]: > please remove two empty lines

Done

> I know we should have a code auto-formatter

the proposed code was after autoformatting, but apparently it is not as strict (PyCharm defaults) as we would like

> but I was hoping to rewrite these scripts to Rust for a long time not to invest time in Python CI)

will be great but in Python is good too

> sigh I wish we had an extra year to dedicate fully to the tech debt.

yep, our common engineering dream :)

File: tools/jsonschema_to_rust_dto_enum_flags.py (hunk 4 above)
Comment (s373r) [reply tools/jsonschema_to_rust_dto_enum_flags.py]: Done... and it's OK, these are helper scripts, so there's no problem :)

**Pull request merged**